### PR TITLE
Add error handling to the account closure flow

### LINKED ISSUE #18406: Add error handling to the account closure flow
This task is to add error handling to the flow for the various types of errors that can be encountered when trying to close the account.

### PR DESCRIPTION
**Pull request by mkevins**
Fixes #18406

### Description

This PR adds error handling to the account closure flow. It depends on FluxC changes brought by this PR: https://github.com/wordpress-mobile/WordPress-Android/pull/18407.

### To test:

Please test using [the steps in the feature branch PR description](https://github.com/wordpress-mobile/WordPress-Android/pull/18412).

PR submission checklist:

See feature branch PR for checklist.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsFragment.kt
Comment (ovitrif) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsFragment.kt]: AFAIK `launchWhenStarted` was deprecated 🤔 .
Can we achieve the same result `repeatOnLifeCycle`?

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/components/DialogErrorUi.kt
Comment (ovitrif) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/components/DialogErrorUi.kt:43]: It's not clear to me what value does it bring to the user if we use different messages for the errors. Would there be cases where users don't end up contacting support because they figured out how to avoid the error by reading the message we display?

**Review by ovitrif — CHANGES_REQUESTED** [review]
I tested the implementation and encountered an unexpected behavior:

I tried to close my account with the username `ovitestin` by:
1. tap `Close account`
2. tap `OK` (the positive button)
3. Not expected: The loading spinner kept spinning forever with no other update on the UI.
4. Then I reinstalled the ap
5. I was logged out, I tried to login and couldn't because apparently "my account is deleted, I should contact support or something, can't understand the whole message because it's trimmed with an ending `…`".

Ps.: The experience I described above is not something the users should encounter IMHO.

It's not very clear to me what we want to achieve with this implementation, but I thought we want to let users close their account. Which in EU is a legal right (the right to be forgotten), so we **must** close the account of a user if requested by them, and we have 30 days to do it.

With that in mind, I don't think users need to see a customised error message that appears not helpful in any way (a few of them could all be "An error occured… contact support to fix it"). But I think they need a working app experience and confirmation the account was closed so they don't try to login again with that account, pretty much what I did, or else they might discover a couple more suboptimal open doors within the app, pretty much like how I discovered these 2 irrelevant error messages for my case, in the onboarding/login flow:

| errMail | nextErrs |
| - | - |

Ps. Thanks for your work @mkevins, things are looking good implementation wise. My concerns are with the solution and the focus which might be consuming too much energy during implementation, while not really providing a reasonable ROI.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/components/DialogErrorUi.kt
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/components/DialogErrorUi.kt:43]: This is a good question. The design goals here are to mirror the iOS implementation, which has these error messages, which I believe were obtained (and perhaps adapted?) from the web experience. In these circumstances, the user wants to close their account, but cannot do so without intervention from support folks (for reasons described in the messages themselves).

> Would there be cases where users don't end up contacting support because they figured out how to avoid the error by reading the message we display?

This is unlikely since they will still need to contact support, afaik.

One good reason for offering a less generic response in case of error is that the user in these cases will continue on to support to complete their desired action. The specific messages can provide better continuity in the experience, and help them reach their goals with less time and effort.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsFragment.kt
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsFragment.kt]: Great idea.. in fact, already done w/ https://github.com/wordpress-mobile/WordPress-Android/commit/d8dea834da7b404f37d767e31e3ba7a4bec2ceeb :grin:  . Since this is to land in a feature branch, I think it's ok get it in along with the tracks PR - but let me know if you think otherwise, and I can cherry-pick it into this branch.

**Comment by mkevins** [comment]
> I tested the implementation and encountered an unexpected behavior:
> 
> I tried to close my account with the username `ovitestin` by:
> 
> 1. tap `Close account`
> 2. tap `OK` (the positive button)
> 3. Not expected: The loading spinner kept spinning forever with no other update on the UI.
> 4. Then I reinstalled the ap
> 5. I was logged out, I tried to login and couldn't because apparently "my account is deleted, I should contact support or something, can't understand the whole message because it's trimmed with an ending `…`".
> 
> Ps.: The experience I described above is not something the users should encounter IMHO.

Hm.. step 3 doesn't sound right - but since I haven't updated all the branches yet, it may be worth testing this once more when it is ready for review (I'll add testing steps).

> It's not very clear to me what we want to achieve with this implementation, but I thought we want to let users close their account. Which in EU is a legal right (the right to be forgotten), so we **must** close the account of a user if requested by them, and we have 30 days to do it.

Can you elaborate on what isn't clear about this implementation? The goal is to let users close their account, which is not only for compliance with EU regulations, but also with upcoming enforcement of Play Store policy.

> With that in mind, I don't think users need to see a customised error message that appears not helpful in any way (a few of them could all be "An error occured… contact support to fix it").

Actually, the error message can be very helpful for the user, and for support folks assisting them. I believe this is the reason why these different error messages exist on web and iOS as well ([where the strings were sourced](https://github.com/wordpress-mobile/WordPress-iOS/blob/b73a00f32b811880262e99b168a58e6a2f229bea/WordPress/Classes/ViewRelated/Me/Account%20Settings/AccountSettingsViewController.swift#L344)).

Consider the case where we provide a generic error message: they user will still have to contact support, and support will have little information to help them troubleshoot why they can't close the account. It will consume more of the user's time, and more time and effort from the support folks. Overall, I think that would actually reduce the value when we can indeed do better.

> But I think they need a working app experience and confirmation the account was closed so they don't try to login again with that account, pretty much what I did, or else they might discover a couple more suboptimal open doors within the app, pretty much like how I discovered these 2 irrelevant error messages for my case, in the onboarding/login flow:

Indeed, the happy path should be a smooth one. That said, once the account is closed, I don't think the user should expect to be able to log in again (the account won't exist anymore). So, I think those error messages are actually relevant.

> Ps. Thanks for your work @mkevins, things are looking good implementation wise. My concerns are with the solution and the focus which might be consuming too much energy during implementation, while not really providing a reasonable ROI.

I'm not entirely sure I understand your concern. Can you elaborate? This task is a compliance task, so the ROI is something that should be given the full weight of consideration with comparison with the case of non-compliance, which is that future updates (that means adding _any_ additional features to the app) would be blocked without it. That said, we do have some time before enforcement - but I think it's a given that non-compliance is a non-starter.

On the focus and energy aspects, do you have an alternative proposal for addressing this task?

**Comment by mkevins** [comment]
I have marked this as ready for review - the code changes can be reviewed in this PR, but please test all flows by [steps in the feature branch PR description](https://github.com/wordpress-mobile/WordPress-Android/pull/18412).

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:59]: The implementation worked as expected for me but I've noticed a compilation error on the `AccountSettingsViewModelTest` test code. Trying to come up with a quick fix for the test I realised that it might be easier to mock the call by using a Use Case (like we do for fetching/pushing account settings). Wdyt?

**Review by ovitrif — APPROVED** [review]
Good work @mkevins 🥇 

The functionality works as expected and the code changes LGTM 🚀 

To fix the compilation issues in `AccountSettingsViewModelTest` you could use this PR:
- https://github.com/wordpress-mobile/WordPress-Android/pull/18450

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt
Comment (mkevins) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:59]: Great idea! I have addressed this (https://github.com/wordpress-mobile/WordPress-Android/pull/18457/commits/93f9741b49c1d1b8572effd0057abc442a70accd), along with adding some unit tests in [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/18457).

**Comment by mkevins** [comment]
Thank you Antonis and Ovi for reviewing and testing!

> Good work @mkevins 1st_place_medal
> 
> The functionality works as expected and the code changes LGTM rocket
> 
> To fix the compilation issues in `AccountSettingsViewModelTest` you could use this PR:
> 
> * [Fix compilation issues in account settings vm tests #18450](https://github.com/wordpress-mobile/WordPress-Android/pull/18450)

:tada: Thank you for fixing this! I've merged it to fix the test build issue. :+1:

**Review by ovitrif — APPROVED** [review]
Reviewed and approved after the merge

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:59]: Thank you Matthew 🙇

**Comment by mkevins** [comment]
Thanks!

**Pull request merged**
